### PR TITLE
Fixes NPM summarizer for legacy licenses key

### PR DESCRIPTION
**Pull request by tmarble**
Handles legacy licenses key as either:
- a string
- an array of strings
- an array of objects (where each type key has the license string)

Resolves: #458 (bis)

Signed-off-by: Tom Marble <tmarble@info9.net>

### CODE REVIEW

File: providers/summary/clearlydefined.js [hunk 1]
```diff
@@ -245,19 +245,39 @@ class ClearlyDescribedSummarizer {
       } else setIfValue(result, 'described.issueTracker', bugs.url || bugs.email)
     }
     // Combine multiple licenses with AND (as it is more restrictive than OR)
+    const relation = ' AND '
     let expression = null  // license expression (as input to SPDX.normalize)
     if (typeof manifest.license === 'string') {
-      expression = manifest.license
+      expression = manifest.license // #1
```
Comment (geneh) [inline providers/summary/clearlydefined.js]: Let's remove these comments

File: providers/summary/clearlydefined.js [hunk 2]
```diff
@@ -245,19 +245,39 @@ class ClearlyDescribedSummarizer {
...
+      expression = manifest.license.type.join(relation) // #4
     } else if (typeof manifest.licenses === 'string') {
-      expression = manifest.licenses // handle legacy NPM 'licenses' key
+      expression = manifest.licenses // #5 handle legacy NPM 'licenses' key
+    } else if (manifest.licenses && (typeof manifest.licenses.type === 'string')) {
+      expression = manifest.licenses.type // #6 handle sub-property 'type'
     } else if (Array.isArray(manifest.licenses)) {
-      expression = manifest.licenses.join(' AND ')
+      // could be an array of strings OR  {type, url} objects
+      const joinLicense = (exp, license) => {
```
Comment (geneh) [inline providers/summary/clearlydefined.js]: This is a lot of conditional code. Could it be something like licenses.map(...).join(' AND ') or at least conditional statements like ": ?"? 
https://docs.clearlydefined.io/contributing-code

> Consistency – A consistently formatted, named, structured, … codebase is easier to work with. Look around at the other code and see how it is structured and formatted.
Minimal – Simplicity and minimalism in the algorithms, data structures and the code itself makes for simpler systems.
Factored – Small, single purpose functions and classes are easier to test and understand.

File: providers/summary/clearlydefined.js [hunk 3]
```diff
@@ -280,22 +291,9 @@ class ClearlyDescribedSummarizer {
...
-    } else if (Array.isArray(manifest.licenses)) {
-      expression = manifest.licenses.join(' OR ')
-    }
-    if (expression) {
-      const licenses = SPDX.normalize(expression)
-      setIfValue(result, 'licensed.declared', licenses)
-    }
+    const expression = this.parseLicenseExpression(manifest, 'OR')
+    if (!expression) return
+    setIfValue(result, 'licensed.declared', SPDX.normalize(expression))
```
Comment (geneh) [inline providers/summary/clearlydefined.js]: Looking at the composer data, it seems like there is no licenses element in the manifest. Has it been mentioned anywhere in the documentation? If not, please revert the changes in this function to the original code.

File: test/summary/clearlydefinedTests.js [hunk 4]
```diff
@@ -271,11 +269,40 @@ describe('ClearlyDescribedSummarizer addNpmData', () => {
...
+    assert.deepEqual(result3, {...expectedResult, licensed: { declared: 'MIT AND Apache-2.0'}})
+
+    let result4 = {}
+    summarizer.addNpmData(result4, { registryData: { manifest: { licenses: licenseArray2 } } }, npmTestCoordinates)
+    assert.deepEqual(result4, {...expectedResult, licensed: { declared: 'MIT AND Apache-2.0'}})
+
+    const licenseArray3 = [ { type: 'MIT'} , { type: 'Apache-2.0'} ]
+    let result5 = {}
+    summarizer.addNpmData(result5, { registryData: { manifest: { licenses: licenseArray3 } } }, npmTestCoordinates)
+    assert.deepEqual(result5, {...expectedResult, licensed: { declared: 'MIT AND Apache-2.0'}})
```
Comment (tmarble) [inline test/summary/clearlydefinedTests.js]: This is where the array of `{type:"", url: "..."}` objects is tested.

File: test/summary/clearlyDefined.js [hunk 5]
```diff
@@ -210,6 +210,22 @@ describe('ClearlyDefined NPM summarizer', () => {
...
+  it('handles a license as an object', () => {
+    const { coordinates, harvested } = setupNpm(null, { type: 'MIT' })
+    const summary = Summarizer().summarize(coordinates, harvested)
+    validate(summary)
+    expect(summary.licensed.declared).to.eq('MIT')
+    expect(summary.described.urls).not.to.be.undefined
+  })
+
+  it('handles a license as an object with an array of strings ', () => {
+    const { coordinates, harvested } = setupNpm(null, { type: ['MIT', 'Apache-2.0'] })
```
[the remark this answers is not in the document]
Comment (tmarble) [reply test/summary/clearlyDefined.js]: There is a test for this case, please see
https://github.com/clearlydefined/service/pull/682/files#r354574099

File: test/summary/clearlydefinedTests.js (hunk 4 above)
Comment (geneh) [reply test/summary/clearlydefinedTests.js]: Perfect, thanks!

File: providers/summary/clearlydefined.js (hunk 3 above)
Comment (geneh) [reply providers/summary/clearlydefined.js]: So I ran a script over all dev & production composer definitions and it looks like they do not have licenses field. There is only the license field. Have you also seen any occurrences of license array with objects and not strings?  For the sake of reducing the complexity, can we go back to the original code for the composer type:
```
if (Array.isArray(manifest.license)) {
  const licenses = SPDX.normalize((manifest.license || []).join(' OR '))
  setIfValue(result, 'licensed.declared', licenses)
}
```

Comment (tmarble) [reply providers/summary/clearlydefined.js]: I just added a fix to not bother with manifest.licenses for composer... Please see if this is acceptable.

File: providers/summary/clearlydefined.js [hunk 6]
```diff
@@ -215,6 +215,32 @@ class ClearlyDescribedSummarizer {
     mergeDefinitions(result, newDefinition, get(result, 'licensed.declared') === 'OTHER')
   }
 
+  parseLicenseExpression(manifest, packageType) {
+    const combineLicenses = (exp, license) => {
+      if (exp) {
+        return exp + ' ' + (packageType==='npm'?'AND':'OR') + ' ' +
```
Comment (geneh) [inline providers/summary/clearlydefined.js]: Could you please enable Prettier & ESLint plugins in VSCode for this repo? It should format everything including spacing between ===, ?, : etc. Just want to make sure everything is consistent.
https://docs.clearlydefined.io/contributing-code

Comment (tmarble) [reply providers/summary/clearlydefined.js]: I think I have fixed the offending lines.. Could you please verify?
